data-prepper-expression/src/main/java/org/opensearch/dataprepper/expression/ParseTreeEvaluatorListener.java

```diff
@@ -0,0 +1,387 @@
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.NOT_EQUAL;
+import static org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser.OR;
+
+/**
+ * @since 1.3
+ * This listener implements {@link DataPrepperExpressionListener} to provide callbacks to handle evaluation of
+ * {@link org.antlr.v4.runtime.tree.ParseTree} representation of an expression while {@link org.antlr.v4.runtime.tree.ParseTreeWalker}
+ * traverses through the {@link org.antlr.v4.runtime.tree.ParseTree}.
+ *
+ * Use case:
+ * ParseTreeWalker walker = new ParseTreeWalker();
+ * ParseTreeEvaluatorListener listener = new ParseTreeEvaluatorListener(...);
+ * listener.initialize(...);
+ * walker.walk(listener, ...);
+ * final Object result = listener.getResult();
+ */
+@Named
+class ParseTreeEvaluatorListener implements DataPrepperExpressionListener {
+    private static final Set<Integer> CONDITIONAL_OPERATOR_TYPES = Set.of(AND, OR);
+    private static final Set<Integer> EQUALITY_OPERATOR_TYPES = Set.of(EQUAL, NOT_EQUAL);
+    private static final Set<Integer> RELATIONAL_OPERATOR_TYPES = Set.of(LT, LTE, GT, GTE);
+
+    private final Map<Integer, Operator<?>> strategy;
+    private final CoercionService coercionService;
+    private Stack<Integer> operatorSymbolStack = new Stack<>();
+    private Stack<Object> operandStack = new Stack<>();
+    private Event context;
+
+    @Inject
+    ParseTreeEvaluatorListener(final List<Operator<?>> operators, final CoercionService coercionService) {
+        this.coercionService = coercionService;
+        strategy = new HashMap<>(operators.size());
+        for (final Operator<?> operator : operators) {
+            strategy.put(operator.getSymbol(), operator);
+        }
+    }
+
+    public void initialize(final Event context) {
+        this.context = context;
+        operatorSymbolStack = new Stack<>();
+        operandStack = new Stack<>();
+    }
+
+    public Object getResult() {
+        if (operandStack.size() != 1 || context == null) {
+            throw new IllegalStateException("The ParseTreeEvaluatorListener has not been initialized or walked through exactly once " +
+                    "after initialization by a ParseTreeWalker.");
+        }
+        return operandStack.peek();
+    }
+
+    @Override
+    public void enterExpression(DataPrepperExpressionParser.ExpressionContext ctx) {
+
+    }
+
+    @Override
+    public void exitExpression(DataPrepperExpressionParser.ExpressionContext ctx) {
+
+    }
+
+    @Override
+    public void enterConditionalExpression(DataPrepperExpressionParser.ConditionalExpressionContext ctx) {
+
+    }
+
+    @Override
+    public void exitConditionalExpression(DataPrepperExpressionParser.ConditionalExpressionContext ctx) {
+        if (operatorSymbolStack.isEmpty() || !CONDITIONAL_OPERATOR_TYPES.contains(operatorSymbolStack.peek())) {
```

Comment:
If we use the concept of the `OperatorSet` enum from my comment above, then all of these `exit` method can be combined. You just need to pass in the `OperatorSet`.

e.g.

```
exitExpression(2, OperatorSet.CONDTIONAL, context);
```

The `exitExpression` method can be common.

You would need to perform the operator validation in `performSingleOperation` after determining the operation.